src/apprt/gtk/portal/OpenURI.zig

```diff
@@ -0,0 +1,400 @@
+        .token = token,
+        .kind = value.kind,
+        .uri = try alloc.dupeZ(u8, value.url),
+    });
+
+    const entry = self.entries.getPtr(token) orelse {
+        return error.PutFailed;
+    };
+
+    self.startCleanupTimer();
+
+    try self.subscribeToResponse(entry);
+    try self.sendRequest(entry);
+}
+
+pub fn deinit(self: *OpenURI) void {
+    const alloc = self.app.core_app.alloc;
+
+    var it = self.entries.valueIterator();
+    while (it.next()) |value| {
+        value.deinit(alloc);
+    }
+
+    self.entries.deinit(alloc);
+}
+
+/// Subscribe to the D-Bus signal that will contain the results of our method
+/// call to the portal. This must be called with the mutex locked.
+fn subscribeToResponse(self: *OpenURI, entry: *Entry) (Allocator.Error || Errors)!void {
+    const alloc = self.app.core_app.alloc;
+
+    if (entry.subscription != null) return;
+
+    const request_path = try portal.getRequestPath(alloc, self.dbus, entry.token);
+    defer alloc.free(request_path);
+
+    entry.subscription = self.dbus.signalSubscribe(
+        null,
+        "org.freedesktop.portal.Request",
+        "Response",
+        request_path,
+        null,
+        .{},
+        responseReceived,
+        self,
+        null,
+    );
+}
+// Unsubscribe to the D-Bus signal that contains the result of the method call.
+// This will prevent a response from being processed multiple times. This must
+// be called when the mutex is locked.
+fn unsubscribeFromResponse(self: *OpenURI, entry: *Entry) void {
+
+    // Unsubscribe from the response signal
+    if (entry.subscription) |subscription| {
+        self.dbus.signalUnsubscribe(subscription);
+        entry.subscription = null;
+    }
+}
+
+/// Send the D-Bus method call to the portal. The mutex must be locked when this
+/// is called.
+fn sendRequest(self: *OpenURI, entry: *Entry) Allocator.Error!void {
+    const alloc = self.app.core_app.alloc;
+
+    const payload = payload: {
+        const builder_type = glib.VariantType.new("(ssa{sv})");
+        defer glib.free(builder_type);
+
+        // Initialize our builder to build up our parameters
+        var builder: glib.VariantBuilder = undefined;
+        builder.init(builder_type);
+
+        // parent window - empty string means we have no window
+        builder.add("s", "");
+
+        // URI to open
+        builder.add("s", entry.uri.ptr);
+
+        // Options
+        {
+            const options = glib.VariantType.new("a{sv}");
+            defer glib.free(options);
+
+            builder.open(options);
+            defer builder.close();
+
+            {
+                const option = glib.VariantType.new("{sv}");
+                defer glib.free(option);
+
+                builder.open(option);
+                defer builder.close();
+
+                builder.add("s", "handle_token");
+
+                const token = try std.fmt.allocPrintZ(alloc, "{x:0<16}", .{entry.token});
```

Comment:
Is token always fixed size (or a predictable max)? Can we get away with a bufAlloc here?

Reply:
Token can be any size we choose as long as it's reasonably certain that it's unique.